In [2]:
import numpy as np
import pandas as pd
import datetime as dt
import xarray as xr
from urllib.error import URLError

import sys
sys.path.append("/home/elilouis/sublimationofsnow/")
import sosutils

import altair as alt
alt.data_transformers.disable_max_rows()
alt.data_transformers.enable('json')

DataTransformerRegistry.enable('json')

In [3]:
sos_download_dir='/data2/elilouis/sublimationofsnow/sosnoqc'
DATE_FORMAT_STR = '%Y%m%d'
start_date = '20230215'; 
end_date = '20230215'; 

datelist = pd.date_range(
    dt.datetime.strptime(start_date, DATE_FORMAT_STR),
    dt.datetime.strptime(end_date, DATE_FORMAT_STR),
    freq='d'
).strftime(DATE_FORMAT_STR).tolist()

VARIABLE_NAMES = [
    # Snow-level temperature arrays (tower D)
    'Tsnow_0_4m_d', 'Tsnow_0_5m_d', 'Tsnow_0_6m_d', 'Tsnow_0_7m_d', 'Tsnow_0_8m_d', 'Tsnow_0_9m_d', 'Tsnow_1_0m_d', 'Tsnow_1_1m_d', 'Tsnow_1_2m_d', 'Tsnow_1_3m_d', 'Tsnow_1_4m_d', 'Tsnow_1_5m_d',

    'Tsnow_0_4m_uw', 'Tsnow_0_5m_uw', 'Tsnow_0_6m_uw', 'Tsnow_0_7m_uw', 'Tsnow_0_8m_uw', 'Tsnow_0_9m_uw', 'Tsnow_1_0m_uw', 'Tsnow_1_1m_uw', 'Tsnow_1_2m_uw', 'Tsnow_1_3m_uw', 'Tsnow_1_4m_uw', 'Tsnow_1_5m_uw',
]

In [4]:
# We make sure that we aren't accessing variables that don't exist in the datasets
# This is necessary because some daily NetCDF files don't have all the expected variables
# (for example because an instrument was down). In that case, we want to add that variable
# to the dataset, filled with nans, which sosutils.merge_datasets_with_different_variables
# handles for us
datasets = []
datasets_safe = []
for date in datelist:
    try:
        ds = xr.open_dataset(sosutils.download_sos_data_day(date, sos_download_dir, cache=True))
    # Some dates are missing
    except URLError:
        print(f"failed on {date}, skipping")
    ds_new = ds[set(ds.data_vars).intersection(VARIABLE_NAMES)]
    datasets.append(ds_new)
    datasets_safe.append(ds_new)
sos_ds = sosutils.merge_datasets_with_different_variables(datasets, dim='time')

Caching...skipping download for 20230215


In [5]:
df = sosutils.get_tidy_dataset(
    sos_ds,
    [
        'Tsnow_0_4m_d',
        'Tsnow_0_5m_d',
        'Tsnow_0_6m_d',
        'Tsnow_0_7m_d',
        'Tsnow_0_8m_d',
        'Tsnow_0_9m_d',
        'Tsnow_1_0m_d',
        'Tsnow_1_1m_d',
        'Tsnow_1_2m_d',
        'Tsnow_1_3m_d',
        'Tsnow_1_4m_d',
        'Tsnow_1_5m_d',

        'Tsnow_0_4m_uw',
        'Tsnow_0_5m_uw',
        'Tsnow_0_6m_uw',
        'Tsnow_0_7m_uw',
        'Tsnow_0_8m_uw',
        'Tsnow_0_9m_uw',
        'Tsnow_1_0m_uw',
        'Tsnow_1_1m_uw',
        'Tsnow_1_2m_uw',
        'Tsnow_1_3m_uw',
        'Tsnow_1_4m_uw',
        'Tsnow_1_5m_uw',
    ]
)

In [6]:
alt.Chart(df).mark_line().encode(
    alt.X("time:T"),
    alt.Y("value:Q"),
    alt.Color("height:O").scale(scheme='rainbow'),
    alt.Facet("tower:N", columns=1),
    tooltip='height'
).properties(width=900, height = 250)

alt.Chart(...)

The 2D heat equation is
$$
\frac{\partial}{\partial t}  T(z,t) = \alpha \frac{\partial^2}{\partial z^2} T(z,t)
$$

We discretize this equation with

$$z_i = i \Delta z$$
$$t_j = j \Delta t$$
$$T(z,t) = T_i^j$$

Using a simple forward differencing scheme to estimate derivatives, we can discretize the above equation and solve for $k^2$.

$$ \alpha =  \frac{T_i^{j+1} - T_i^{j}}{\Delta t} \Big( \frac{\Delta z^2}{T_{i+1}^{j} - 2 T_{i}^{j} + T_{i-1}^{j}} \Big)$$

Note that this scheme is numerically stable when 

$$ \Delta t \leq \frac{\Delta x^2}{4 k^2}$$

The solution for $k^2$ allows us to solve for k for every data point in the (z,t) space that isn't on the edge of the (z,t) grid.

In our dataset, $\Delta t$ is 300 seconds and $\Delta x$ is 0.1 meters. We now transform our dataset to be in a 2-D array.

In [7]:
Tsnow_d = np.array([
    sos_ds['Tsnow_0_4m_d'].values,
    sos_ds['Tsnow_0_5m_d'].values,
    sos_ds['Tsnow_0_6m_d'].values,
    sos_ds['Tsnow_0_7m_d'].values,
    sos_ds['Tsnow_0_8m_d'].values,
    sos_ds['Tsnow_0_9m_d'].values,
    sos_ds['Tsnow_1_0m_d'].values,
    sos_ds['Tsnow_1_1m_d'].values,
    # sos_ds['Tsnow_1_2m_d'].values,
    # sos_ds['Tsnow_1_3m_d'].values,
    # sos_ds['Tsnow_1_4m_d'].values,
    # sos_ds['Tsnow_1_5m_d'].values
])

Tsnow_uw = np.array([
    sos_ds['Tsnow_0_4m_uw'].values,
    sos_ds['Tsnow_0_5m_uw'].values,
    sos_ds['Tsnow_0_6m_uw'].values,
    sos_ds['Tsnow_0_7m_uw'].values,
    sos_ds['Tsnow_0_8m_uw'].values,
    sos_ds['Tsnow_0_9m_uw'].values,
    sos_ds['Tsnow_1_0m_uw'].values,
    sos_ds['Tsnow_1_1m_uw'].values,
    sos_ds['Tsnow_1_2m_uw'].values,
    sos_ds['Tsnow_1_3m_uw'].values,
    # sos_ds['Tsnow_1_4m_uw'].values,
    # sos_ds['Tsnow_1_5m_uw'].values
])

This grid allows us to access our data with $i$ and $j$ indices like so:

In [8]:
i = 0; j = 0
print(Tsnow_d[i][j])
i = 0; j = 1
print(Tsnow_d[i][j])
i = 1; j = 0
print(Tsnow_d[i][j])
i = 1; j = 1
print(Tsnow_d[i][j])

-3.5763922
-3.5760107
-4.2438574
-4.2434115


In [9]:
Tsnow_d.shape

(8, 288)

Now we can step through our space and time grid. We have 12 spatial measurements each with 288 time steps.

$i \quad \epsilon \quad [0, 11]$

$j \quad \epsilon \quad [0, 287]$

where $i$ corresponds to measurements between 0.4 and 1.5 meters, at 0.1 meter increments.

Using our equation for $k^2$ above, repeated here

$$ \alpha =  \frac{T_i^{j+1} - T_i^{j}}{\Delta t} \Big( \frac{\Delta z^2}{T_{i+1}^{j} - 2 T_{i}^{j} + T_{i-1}^{j}} \Big)$$


We see that we can only calculate k^2 with the following subset of $i$ and $j$

$i \quad \epsilon \quad (0, 11)$

$j \quad \epsilon \quad [0, 287)$

Let's step through this data and calculate k^2! But first let's convert to Kelvin :).

In [10]:
Tsnow_d = Tsnow_d + 273.15
Tsnow_uw = Tsnow_uw + 273.15

In [11]:
delta_z = 0.1
delta_t = 300

i_range_d = range(1,7)
i_range_uw = range(0,9)

alpha_d = np.zeros_like(Tsnow_d)
alpha_uw = np.zeros_like(Tsnow_uw)

for i in i_range_d:
    for j in range(0, 287):
        alpha_d[i][j] = (
            (Tsnow_d[i][j+1] - Tsnow_d[i][j])/delta_t
        )*(
            delta_z**2 / ( Tsnow_d[i+1][j] - 2*Tsnow_d[i][j] + Tsnow_d[i-1][j])
        )

for i in i_range_uw:
    for j in range(0, 287):
        alpha_uw[i][j] = (
            (Tsnow_uw[i][j+1] - Tsnow_uw[i][j])/delta_t
        )*(
            delta_z**2 / ( Tsnow_uw[i+1][j] - 2*Tsnow_uw[i][j] + Tsnow_uw[i-1][j])
        )

In [12]:
np.nanmean(alpha_d), np.nanmedian(alpha_d), np.max(alpha_d), np.min(alpha_d)

(4.1044865e-07, 0.0, 0.00024222222, -0.00022522523)

In [13]:
np.nanmean(alpha_uw), np.nanmedian(alpha_uw), np.max(alpha_uw), np.min(alpha_uw)

(1.2061288e-06, 0.0, 0.0017897435, -0.0014098039)

In [34]:
src = pd.DataFrame({
    'thermal diffusivity (tower d)': alpha_d.flatten(),
    'thermal diffusivity (tower uw)': alpha_uw.flatten()
})
alt.Chart(src).mark_bar().transform_fold(
    ['thermal diffusivity (tower d)', 'thermal diffusivity (tower uw)']
).encode(
    alt.X("value:Q").title("α"),
    alt.Y("count():Q"),
    alt.Facet("key:N").title(None)
).properties(height=150)

<VegaLite 5 object>

If you see this message, it means the renderer has not been properly enabled
for the frontend that you are using. For more information, see
https://altair-viz.github.io/user_guide/display_frontends.html#troubleshooting


In [29]:
src

time  thermal diffusivity (tower d)  \
0   2023-02-15 00:02:30                  -3.883607e-08   
1   2023-02-15 00:07:30                   1.150351e-07   
2   2023-02-15 00:12:30                   5.211264e-08   
3   2023-02-15 00:17:30                   1.216424e-07   
4   2023-02-15 00:22:30                  -1.480078e-07   
..                  ...                            ...   
283 2023-02-15 23:37:30                   1.637137e-07   
284 2023-02-15 23:42:30                   4.743436e-07   
285 2023-02-15 23:47:30                   5.532177e-06   
286 2023-02-15 23:52:30                  -9.798421e-06   
287 2023-02-15 23:57:30                   0.000000e+00   

     thermal diffusivity (tower uw)  
0                     -2.093634e-07  
1                     -1.126318e-07  
2                     -1.273473e-07  
3                     -1.163467e-07  
4                     -1.526793e-07  
..                              ...  
283                   -2.450575e-06  
284                   -5.431519e-07  
285                   -1.046003e-05  
286                   -8.664067e-06  
287                    0.000000e+00  

[288 rows x 3 columns]

In [37]:

src = pd.DataFrame({
    'time': df.time.unique(),
    'thermal diffusivity (tower d)': alpha_d.mean(axis=0),
    'thermal diffusivity (tower uw)': alpha_uw.mean(axis=0)
})
alt.Chart(src).mark_line().transform_window(
    frame=[-12, 12],
    avg = 'mean(thermal diffusivity (tower uw))',
).encode(
    alt.X("time:T"),
    alt.Y("avg:Q").title("α"),
).properties(height=150)

alt.Chart(...)